src/main/java/com/thoughtworks/aceleradora/controladores/MinhaListaControlador.java

```diff
@@ -39,5 +48,38 @@ public String salvarLista(MinhaLista lista) {
         return "minhaLista/cadastro";
     }
 
+    @GetMapping("/editar-lista/{id}")
+    public String pegaLista(Model modelo, @PathVariable("id") Long id) {
+        Optional<MinhaLista> lista = minhaListaServico.encontraUm(id);
+
+        modelo.addAttribute("lista", lista.get());
+
+        return "minhaLista/editar";
+    }
+
+    @PostMapping("/editar-lista/{id}")
+    public String salvarEdicao(MinhaLista minhaLista, @PathVariable("id") Long id) {
+        for(int i = 0; i < minhaLista.getProdutos().size(); i++) {
+            minhaListaServico.atualizar(id, minhaLista.getProdutos().get(i).getId());
+        }
+
+        return "minhaLista/cadastro";
+    }
+
+    @PostMapping("/editar-lista/{id}/remover-item")
+    public String removerItem(MinhaLista ListaDoFronte, @PathVariable("id") Long id) {
+        Optional<MinhaLista> listaDoBanco = minhaListaServico.encontraUm(id);
+        List<Produto> devoDeletar = new ArrayList<>();
```

Comment:
O uso de Streams deixou o código mais complicado do que ele deveria ser. O uso do bom e velho `for` ou `foreach` deixaria as coisas mais simples, creio:

```java
Optional<MinhaLista> listaRegistrada = minhaListaServico.encontraUm(id);

List<Produto> produtosParaSeremRemovidos = new ArrayList<>();

//se existir o registro no banco de dados...
if (listaRegistrada.isPresent()) {
  List<Produto> produtosCadastrados = listaRegistrada.get().getProdutos();

  for(Produto produtoDoFront : listaDoFronte.getProdutos()) {
    if (!produtosCadastrados.contains(produtoDoFront)) {
      produtosParaSeremRemovidos.add(produtoDoFront);
    }
  }
  
  //Atualiza copia em memoria da lista cadastrada, removendo os produtos deselecionados
  listaRegistrada.get().getProdutos().removeAll(produtosParaSeremRemovidos);

  // Atualiza a lista no banco, alterando os produtos selecionados
  minhaListaServico.salvar(listaRegistrada.get());
}

```

Tendo dito isso, independente da forma como vamos seguir essa implementação, toda essa lógica deveria estar no serviço, não no controller. O serviço deve saber como interagir com registros de banco de dados, etc. Isso não é papel do controlador.

Reply:
Nós utilizamos a mesma orientação, mas alteramos a ordem para que ele pudesse funcionar: ao invés de pegar do front e comparar com o banco, pegamos do banco primeiro para comparar com o front, e assim, funcionou.